docs(component-store): add usage for local ui state (#2675)
ngrx/platform: projects/ngrx.io/content/guide/component-store/usage.md

The types of state that developers typically deal with in applications are:
* **Server/Backend(s) State**. This is the ultimate source of truth of all the data.
* **Persisted State**. The "snapshots" of backend data transferred *to* and *from* application. E.g. Movies data passed as a JSON response, or user's rating for a particular Movie passed as an update request.
* **URL State**. The state of the URL itself. Depending on which URL use navigates to, the app would open specific pages and thus might request for *Persisted State*.
* **Client State**. The state within application that is not persisted to the backend. E.g. The info about which Tab is open in the application.
* **Local UI State**. The state within a component itself. E.g. `isEnabled` toggle state of Toggle Component.

<figure>
  <img src="generated/images/guide/component-store/types-of-state.png" alt="Types of state in a typical SPA" width="100%" height="100%" />
</figure>

There are more types of states, but these are the most important ones in the context of state management.

<div class="alert is-important">

Synchronizing these states is one of the most complex tasks that developers have to solve.

</div>

> Here is a small example to demonstrate how even a simple task might involve all of them:

Reviewer: 👍

Author: 😃

1. The user opens the page at a specific URL, "https://TheBestMoviesOfAllTimeEver.com/favorites". That changes the ***URL State***. 
1. The URL had a path for a specific tab, "favorites". That selection becomes part of the ***Client State***. 
1. This results in API calls to the backend to get the data of the movies that the user marked as "favorites". We receive a snapshot of ***Persisted State***.
1. The Toggle Component that lives next to the *"is favorite"* label is turned ON. The "ON" state is derived from the data that the application received and passed to the Toggle Component through `@Input() isEnabled: boolean`. The component itself is not aware of *Persisted State* or what it even means to be ON in the context of the rest of the application. All it knows is that it needs to be visually displayed as ON. The `isEnabled` state is ***Local UI State***.
1. The user might decide that this movie is no longer their favorite and would click the Toggle button to turn it OFF. The *Local UI State* of the component is then changed, the `@Output() changed` event is emitted and picked up by a container component which would then call the backend to update the *Persisted State*.

> This was a very simple example. Typically developers have to solve many problems that are more interconnected. What if the user is not logged in? Should we wait until the new favorite state is persisted to the backend and only then show disabled state or should we do this optimistically? and so on.

Author: I'm trying to highlight that many problems are **more interconnected** that the provided example.

<div class="alert is-helpful">

Understanding these types of state helps us define our usage of ComponentStore.

</div>

## Use Case 1: Local UI State

### Example 1: ComponentStore as part of component

The simplest example usage of `ComponentStore` is **reactive *Local UI State***.

> One of the ways to improve the performance of the application is to use the `OnPush` change detection strategy. However, contrary to the popular belief, we do not always need to tell Angular's change detection to `markForCheck()` or `detectChanges()` (or the Angular Ivy alternatives). As pointed out in [this article on change detection](https://indepth.dev/the-last-guide-for-angular-change-detection-youll-ever-need/), if the event originates from the component itself, it will be dirty checked.

Author: Yeah, will move into the `is-helpful` block

> This means that common presentational (aka dumb) components which interact with the rest of the application with Input(s)/Output(s) might not need to be overcomplicated with reactive state even though we did it to the Toggle Component mentioned above.

Reviewer 2: do not to be?

> This means that common presentational (aka dumb) components which interact with the rest of the application with Input(s)/Output(s) might not need to be overcomplicated with reactive state even though we did it to the Toggle Component mentioned above.

Reviewer: Thanks Lars.

> This means that common presentational (aka dumb) components which interact with the rest of the application with Input(s)/Output(s) might not need to be overcomplicated with reactive state even though we did it to the Toggle Component mentioned above.

Author: that -> which -> that after a few reviews 😀

`@Input` here is a setter function that passes the value to the `setChecked` updater.

When a user clicks the toggle, instead of calling the same updater directly, the `onChangeEvent` effect is called. This is done because we also need to have the side-effect of `event.stopPropagation()` and only after that the `setChecked` updater is called with the value of the input element.

<code-example
  path="component-store-slide-toggle/src/app/slide-toggle.component.ts"
  region="updater"></code-example>

#### Step 3. Reading the state

Finally, the state is aggregated with selectors into two properties:
* `vm$` property collects all the data needed for the template - this is *ViewModel* of `SlideToggleComponent`. 
* `change` is the `@Output` of `SlideToggleComponent`. Instead of creating `EventEmitter` and push the data through it, here the output is connected to the Observable source directly.

<code-example
  path="component-store-slide-toggle/src/app/slide-toggle.component.ts"
  region="selector"></code-example>

This example does not have a lot of business logic, however it is still fully reactive.

> #### Zoneless

Author: We now filled these details when Zoneless is mentioned for the first time, so yeah, I'll remove it from here.

In the template of `SlideToggleComponent` we used `async` pipe to unwrap the Observable, however in the Zoneless environment it will not work. In such cases, [ngrxPush pipe](guide/component/push) can be used instead. 

### Example 2: Service extending ComponentStore

`SlideToggleComponent` is a fairly simple component and having `ComponentStore` within the component itself is still manageable. When components takes more Inputs and/or has more events within its template, it becomes larger and harder to read/maintain.

Extracting the business logic of a component into a separate Service also helps reduce the cognitive load while reading the components code.

<div class="alert is-important">

A Service that extends ComponentStore and contains business logic of the component brings many advantages. **This is also the recommended usage of ComponentStore**.

</div>

> `ComponentStore` was designed with such approach in mind. The main APIs of `ComponentStore` (`updater`, `effect` and `select`) are meant to wrap the **HOW** state is changed, extracted or effected, and then provided with inputs.

Author: > What is meant with `effected`?
state is changed (with updaters), extracted (with selectors) or effected (with effects)

> To me `and then provided with inputs` is confusing, what do we mean with it here?
I agree the wording here should be better. What I meant is that these updaters and effects can be called with arguments in Component itself - these arguments are the inputs into the ComponentStore.

I'll try to re-phrase it better

Below are the two examples of re-implemented [Paginator component](https://material.angular.io/components/paginator/overview) from Angular Material library of UI components. These re-implementations are very functionally close alternatives.

> Here's the source code of the [Material's paginator.ts](https://github.com/angular/components/blob/23d3c216c65b327e0acfb48b53302b8fda326e7f/src/material/paginator/paginator.ts#L112) for the reference.

Reviewer: ```suggestion
Here's the source code of the [Material's paginator.ts](https://github.com/angular/components/blob/23d3c216c65b327e0acfb48b53302b8fda326e7f/src/material/paginator/paginator.ts#L112) as a reference.
```

What we can see is that while the *"PaginatorComponent providing ComponentStore"* example already makes the component a lot smaller, reactive, removes `this._changeDetectorRef.markForCheck()` and organizes it into distinct "read"/"write"/"effect" buckets, it still could be harder to read. *"PaginatorComponent with PaginatorStore"* example adds readability and further improves testability of behaviors and business logic.

<div class="alert is-helpful">

You can see the examples at StackBlitz:
* "PaginatorComponent providing ComponentStore" <live-example name="component-store-paginator" noDownload></live-example>
* "PaginatorComponent with PaginatorStore Service" <live-example name="component-store-paginator-service" noDownload></live-example>

</div>

> <code-tabs linenums="true">

Reviewer: Nit, I find this codeblock too big to be added to the docs.
It also makes it hard to scroll to the bottom of the page.

Author: I want to make it easier to compare these two side-by-side. I'll extract the init state into PaginatorStore - that should make it a bit smaller. (was planning to do it anyway :) )

Reviewer: Thanks Alex!

  </code-pane>
  <code-pane
    header="paginator.store.ts"
    path="component-store-paginator-service/src/app/paginator.store.ts">
  </code-pane>
</code-tabs>

#### Updating the state

With `ComponentStore` extracted into `PaginatorStore`, the developer is now using updaters and effects to update the state. Effects are used to perform additional validation and get extra information from sources with derived data (i.e. selectors).

<code-example
  path="component-store-paginator-service/src/app/paginator.component.ts"
  region="updating-state"></code-example>

#### Reading the state

`PaginatorStore` exposes the two properties: `vm$` for an aggregated *ViewModel* to be used in the template and `page$` that would emit whenever data aggregated from a `PageEvent` changes.

<code-tabs>
  <code-pane
    header="paginator.component.ts"
    path="component-store-paginator-service/src/app/paginator.component.ts"
    region="selectors"
    >
  </code-pane>
  <code-pane
    header="paginator.store.ts"
    path="component-store-paginator-service/src/app/paginator.store.ts"
    region="selectors">
  </code-pane>
</code-tabs>

<div class="alert is-helpful">

> `page$` would emit `PageEvent` when page size or page index changes, however in some cases updater would update both of these properties at the same time. To avoid "intermediary" emissions, `page$` selector is using **`{debounce: true}`** configuration. This schedules a microtask, which allows the state to "settle", meaning all the synchronous changes to happen, and then it would emit a value. All of this happens before Angular's Change Detection runs. More about debouncing can be found in [selector section](guide/component-store/read#debounce-selectors).

Author: You are right. I'll shorten it here.

</div>

### Local UI State patterns

Components that use `ComponentStore` for managing **Local UI State** are frequently calling `updater`s directly.

Effects can also be used when:
* side effects are required (e.g. `event.stopPropagation()`)
* derived data (from selectors) is needed to influence the new state
* they are orchestrating a number of well-defined updaters

> The last point can sometimes be refactored into another `updater`. Use your best judgement.

Author: 😮

`@Output()`s and derived data are **reacting** to these state changes and are generated using `selector`s.

A *ViewModel* for the component is also composed from `selector`s.

## Use Case 2: Complex Common Component State

## Use Case 3: Complex Independent Component Store

> ## Use Case 4: Application-wide Singleton Service

Author: I wanted to point out that `Local UI State` is not the only use case. I'm ok with these being empty and docs being published - that signals that these would be added with time, and readers should check back.
I can make it clearer with "The use case is in the process of being written - check back later" for them.
WDYT?

> ## Use Case 4: Application-wide Singleton Service

Reviewer: That's good for me 🙂

* **Persisted State**. The "snapshots" of backend data transferred *to* and *from* application. E.g. Movies data passed as a JSON response, or user's rating for a particular Movie passed as an update request.
* **URL State**. The state of the URL itself. Depending on which URL use navigates to, the app would open specific pages and thus might request for *Persisted State*.
* **Client State**. The state within application that is not persisted to the backend. E.g. The info about which Tab is open in the application.
* **Local UI State**. The state within a component itself. E.g. `isEnabled` toggle state of Toggle Component.

<figure>
  <img src="generated/images/guide/component-store/types-of-state.png" alt="Types of state in a typical SPA" width="100%" height="100%" />
</figure>

There are more types of states, but these are the most important ones in the context of state management.

<div class="alert is-important">

Synchronizing these states is one of the most complex tasks that developers have to solve.

</div>

Here is a small example to demonstrate how even a simple task might involve all of them:

> 1. The user opens the page at specific URL - "https://TheBestMoviesOfAllTimeEver.com/favorites" - that changes ***URL State***. 

Reviewer 2: ```suggestion
1. The user opens the page at a specific URL, "https://TheBestMoviesOfAllTimeEver.com/favorites". That changes the ***URL State***. 
```

> 1. URL had a path for a specific tab - "favorites" - that selection becomes part of ***Client State***. 

Reviewer 2: ```suggestion
1. The URL had a path for a specific tab, "favorites". That selection becomes part of the ***Client State***. 
```

> 1. Now, that results in API calls to the backend to get the data of the movies that user marked as "favorites" - we get the snapshot of ***Persisted State***.

Reviewer 2: ```suggestion
1. This results in API calls to the backend to get the data of the movies that the user marked as "favorites". We receive a snapshot of ***Persisted State***.
```

> 1. The Toggle Component that lives next to *"is favorite"* label is turned ON. The "ON" state is derived from the data that application received and passed to Toggle Component through `@Input() isEnabled: boolean`. The component itself is not aware of *Persisted State*, or what it even means to be ON in the context of the rest of the application. All it knows that it needs to be visually displayed as ON - `isEnabled` state is ***Local UI State***.

Reviewer 2: ```suggestion
1. The Toggle Component that lives next to the *"is favorite"* label is turned ON. The "ON" state is derived from the data that the application received and passed to the Toggle Component through `@Input() isEnabled: boolean`. The component itself is not aware of *Persisted State* or what it even means to be ON in the context of the rest of the application. All it knows is that it needs to be visually displayed as ON. The `isEnabled` state is ***Local UI State***.
```

> 1. The user might decided that this movie is no longer their favorite, and would click the Toggle button. The *Local UI State* of the component is then changed, the `@Output() changed` event is emitted and picked up by the container component, which then would call the backend to update the new *Persisted State*. 

Reviewer 2: ```suggestion
1. The user might decide that this movie is no longer their favorite and would click the Toggle button to turn it OFF. The *Local UI State* of the component is then changed, the `@Output() changed` event is emitted and picked up by a container component which would then call the backend to update the *Persisted State*.
```

> This was a very simple example, typically developers have to solve many more interconnected problems. What if the user is not logged in? Should we wait until the new favorite state is persisted to the backend and only then show disabled state, or should we do this optimistically? and etc.

Reviewer 2: ```suggestion
This was a very simple example. Typically developers have to solve many problems that are more interconnected. What if the user is not logged in? Should we wait until the new favorite state is persisted to the backend and only then show disabled state or should we do this optimistically? and so on.
```

<div class="alert is-helpful">

> The understanding of these state would help us define the usage of ComponentStore.

Reviewer 2: ```suggestion
Understanding these types of state helps us define our usage of ComponentStore.
```

</div>

## Use Case 1: Local UI State

### Example 1: ComponentStore as part of component

> The simplest example of the usage of `ComponentStore` would be the **reactive *Local UI State***.

Reviewer 2: ```suggestion
The simplest example usage of `ComponentStore` is **reactive *Local UI State***.
```

> One of the ways to improve the performance of the application is to use `OnPush` change detection strategy. However, contrary to the popular belief, we do not always need to tell Angular's change detection to `markForCheck()` or `detectChanges()` (or the Ivy's alternatives). As pointed in [the article on change detection](https://indepth.dev/the-last-guide-for-angular-change-detection-youll-ever-need/), if the event originates from the component it would be checked.

Reviewer 2: ```suggestion
One of the ways to improve the performance of the application is to use the `OnPush` change detection strategy. However, contrary to the popular belief, we do not always need to tell Angular's change detection to `markForCheck()` or `detectChanges()` (or the Angular Ivy alternatives). As pointed out in [this article on change detection](https://indepth.dev/the-last-guide-for-angular-change-detection-youll-ever-need/), if the event originates from the component itself, it will be dirty checked.
```

> That means that presentational (aka dumb) common components, that interact with the rest of the application with Input(s)/Output(s) might not need to be overcomplicated with reactive state - just like the Toggle Component mentioned above.

Reviewer 2: ```suggestion
This means that common presentational (aka dumb) components which interact with the rest of the application with Input(s)/Output(s) might not need to be overcomplicated with reactive state even though we did it to the Toggle Component mentioned above.
```

Having said that, in most cases making *Local UI State* **reactive** is beneficial:

> * Zoneless application - where `async` pipe can be easily substituted with Zoneless alternative

Reviewer 2: ```suggestion
* For Zoneless application so that the `async` pipe can easily be substituted with a Zoneless alternative such as [ngrxPush pipe](guide/component/push)
```

> * Components with non-trivial business logic - reactivity can organize the state better, clearly separating actual state from derived values, identify side-effects.

Reviewer 2: ```suggestion
* For components with non-trivial business logic, reactivity can organize the state better by clearly separating actual state from derived values and identifying side-effects.
```

ComponentStore is not the only reactive *Local UI State* holder - sometimes `FormControl`s are good enough. They contain the state and they have reactive APIs.

> Here's the simplified `SlideToggleComponent` example that uses `ComponentStore` for *Local UI State*. In this example, the `ComponentStore` is provided directly to the component itself, which might not be the best choice for most of the use cases of `ComponentStore`.

Reviewer 2: ```suggestion
Here's the simplified `SlideToggleComponent` example which uses `ComponentStore` for *Local UI State*. In this example, the `ComponentStore` is provided directly by the component itself, which might not be the best choice for most of the use cases of `ComponentStore`. Instead, consider a service that `extends ComponentStore`.
```

<div class="alert is-helpful">

> You can see the full example at Stackblitz: <live-example name="component-store-slide-toggle"></live-example>

Reviewer 2: ```suggestion
You can see the full example at StackBlitz: <live-example name="component-store-slide-toggle"></live-example>
```

</div>

`ComponentStore` was designed with such approach in mind. The main APIs of `ComponentStore` (`updater`, `effect` and `select`) are meant to wrap the **HOW** state is changed, extracted or effected, and then provided with inputs.

Below are the two examples of re-implemented [Paginator component](https://material.angular.io/components/paginator/overview) from Angular Material library of UI components. These re-implementations are very functionally close alternatives.

Here's the source code of the [Material's paginator.ts](https://github.com/angular/components/blob/23d3c216c65b327e0acfb48b53302b8fda326e7f/src/material/paginator/paginator.ts#L112) for the reference.

What we can see is that while *"PaginatorComponent providing ComponentStore"* example already makes component a lot smaller, reactive, removes `this._changeDetectorRef.markForCheck()` and organizes it into distinct "read"/"write"/"effect" buckets, it still could be harder to read. *"PaginatorComponent with PaginatorStore"* example adds readability and further improves testability of behaviors and business logic.

<div class="alert is-helpful">

> You can see the examples at Stackblitz:

Reviewer 2: ```suggestion
You can see the examples at StackBlitz:
```

<code-tabs>
  <code-pane
    header="paginator.component.ts"
    path="component-store-paginator-service/src/app/paginator.component.ts"
    region="selectors"
    >
  </code-pane>
  <code-pane
    header="paginator.store.ts"
    path="component-store-paginator-service/src/app/paginator.store.ts"
    region="selectors">
  </code-pane>
</code-tabs>

<div class="alert is-helpful">

`page$` would emit `PageEvent` when page size or page index changes, however in some cases updater would update both of these properties at the same time. To avoid "intermediary" emissions, `page$` selector is using **`{debounce: true}`** configuration. This schedules a microtask, which allows the state to "settle", meaning all the synchronous changes to happen, and then it would emit a value. All of this happens before Angular's Change Detection runs. More about debouncing can be found in [selector section](guide/component-store/read#debounce-selectors).

</div>

### Local UI State patterns

Components that use `ComponentStore` for managing **Local UI State** are frequently calling `updater`s directly.

> Effects can be also called when:

Reviewer 2: Suggestion 1
```suggestion
Effects can also be called when:
```

Suggestion 2
```suggestion
Effects can also be used when:
```

* side effects are required (e.g. `event.stopPropagation()`)
* derived data (from selectors) is needed to influence the new state
* they are orchestrating a number of well-defined updaters

> The last point sometimes could be refactored into `updater`. Use your best judgement.

Reviewer 2: Not exactly sure what you mean by refactoring orchestration of multiple updaters into a single updater.

Suggestion 1
```suggestion
The last point can sometimes be refactored into an aggregate `updater`. Use your best judgement.
```

Suggestion 2
```suggestion
The last point can sometimes be refactored into a single `updater`. Use your best judgement.
```

Suggestion 3
```suggestion
The last point can sometimes be refactored into another `updater`. Use your best judgement.
```

Author: The 3-rd option is what I meant :)

> `@Output()`s and derived data are **reacting** on these state changes and are generated with `selector`s.

Reviewer 2: ```suggestion
`@Output()`s and derived data are **reacting** to these state changes and are generated using `selector`s.
```

> *ViewModel* for the component is collected with `selector` as well.

Reviewer 2: ```suggestion
A *ViewModel* for the component is also composed from `selector`s.
```
